src/lib/classes/QScrollObserver.svelte.ts

```diff
@@ -0,0 +1,109 @@
+type Direction = "up" | "right" | "down" | "left";
+
+type WindowScrollType = "scrollX" | "scrollY";
+type DocumentScrollType = "scrollLeft" | "scrollTop";
+
+type ScrollType = {
+  window: WindowScrollType;
+  document: DocumentScrollType;
+};
+
+/**
+ * A scroll observer utility.
+ * @example
+ * ```js
+ * // Inside the <script> tag
+ * const scroll = new QScrollObserver();
+ *
+ * $inspect(scroll.direction)
+ * ```
+ *
+ * @param horizontal - If set to `true`, observe horizontal scrolling rather than vertical
+ * @param debounce - Time in milliseconds between each observation update
+ */
+export default class QScrollObserver {
+  private scrollType: ScrollType = {
+    window: "scrollY",
+    document: "scrollTop",
+  };
+  private lastScroll: number;
+  private horizontal: boolean = false;
+  private clearTimer: (() => void) | null = null;
+
+  direction = $state<Direction>("down");
+  changedDirection = $state(false);
+  delta = $state(0);
+  inflectionPoint = $state(0);
+  position = $state(0);
+
+  constructor(horizontal = false, debounce = 250) {
+    this.horizontal = horizontal;
+
+    this.scrollType = horizontal
+      ? { window: "scrollX", document: "scrollLeft" }
+      : { window: "scrollY", document: "scrollTop" };
+
+    this.lastScroll = 0;
+    this.position = 0;
+    this.inflectionPoint = 0;
+    this.direction = horizontal ? "right" : "down";
+
+    const handler = (e: Event) => {
+      const target = e.target as HTMLDivElement | null;
+      if (!target || target.parentElement?.classList.contains("q-layout")) {
+        return;
+      }
+
+      if (!debounce) {
```

Comment:
I find this whole block a bit hard to grasp. Can we separate the debounce logic from this classes' logic?

Reply:
I can try but it'll be really annoying to do because this uses the class' functions. I guess I could pull this out with a `.bind(this, debounce, target)` but that'd be a hassle just for some debounce logic

Reply:
I could separate it in a class function though